[TF FE]: Support complex tensors for Prod operations

hub-bla opened the pull request:
### Details:
- Fixed fifth case of `atan2` implementation (it returned `pi/2` instead of `-pi/2`).
- Moved `atan2` to `utils`.
- Create helper functions for converting complex number from rectangular to polar form and the other way around.
- Support complex tensors for `Prod` operations + unit tests.

### Tickets:
 - [None](https://github.com/openvinotoolkit/openvino/issues/23233)


### Resources used:
 - https://math.stackexchange.com/questions/1938894/imaginary-part-of-a-product-of-n-complex-numbers
 - https://en.m.wikipedia.org/wiki/Euler%27s_formula

rkazants commented on tests/layer_tests/tensorflow_tests/test_tf_ReduceArithmeticOps.py:
```diff
@@ -50,3 +50,51 @@ def test_reduce(self, params, operation, keep_dims, ie_device, precision, ir_ver
+        tf.compat.v1.reset_default_graph()
+        with tf.compat.v1.Session() as sess:
+            param_real = tf.compat.v1.placeholder(tf.float32, shape, 'param_real')
+            param_imag = tf.compat.v1.placeholder(tf.float32,shape, 'param_imag')
```
```suggestion
            param_imag = tf.compat.v1.placeholder(tf.float32, shape, 'param_imag')
```

rkazants commented:
build_jenkins

hub-bla commented:
@rkazants I forgot to add space in a cpp file which caused clang_format error. Now it should be fine

hub-bla commented:
Am I allowed to alter `REDUCE_PROD` here
https://github.com/openvinotoolkit/openvino/blob/102e8754440aec4ff5712bccdff009bf7ee88605/src/frontends/tensorflow_lite/src/op_table.cpp#L142
or move translation back to `translate_direct_reduce_op`?

rkazants reviewed (approved):
Great job, I do like it.

rkazants commented:
build_jenkins

hub-bla commented:
> Am I allowed to alter `REDUCE_PROD` here
> 
> https://github.com/openvinotoolkit/openvino/blob/102e8754440aec4ff5712bccdff009bf7ee88605/src/frontends/tensorflow_lite/src/op_table.cpp#L142
> 
> 
> or move translation back to `translate_direct_reduce_op`?

Because of that test in tf_lite is failing so it won't pass now

rkazants commented:
> > Am I allowed to alter `REDUCE_PROD` here
> > https://github.com/openvinotoolkit/openvino/blob/102e8754440aec4ff5712bccdff009bf7ee88605/src/frontends/tensorflow_lite/src/op_table.cpp#L142
> > 
> > or move translation back to `translate_direct_reduce_op`?
> 
> Because test in tf_lite is failing it won't pass now. Let me know what should I do with it

please use `translate_prod_op` translator in TFLite since it is common for both FEs.
btw, there is an failure in TF layer tests due to accuracy error. Please investigate.

Best regards,
Roman

rkazants commented on tests/layer_tests/tensorflow_tests/test_tf_ReduceArithmeticOps.py:
```diff
@@ -50,3 +50,51 @@ def test_reduce(self, params, operation, keep_dims, ie_device, precision, ir_ver
+        assert 'param_imag:0' in inputs_info, "Test error: inputs_info must contain `param_imag`"
+        x_shape = inputs_info['param_real:0']
+        inputs_data = {}
+        inputs_data['param_real:0'] = np.random.randint(-10, 10, x_shape).astype(np.float32)
```
please use `rng` with static seed to avoid sporadic issue. Check other tests with it.

rkazants commented:
> > > Am I allowed to alter `REDUCE_PROD` here
> > > https://github.com/openvinotoolkit/openvino/blob/102e8754440aec4ff5712bccdff009bf7ee88605/src/frontends/tensorflow_lite/src/op_table.cpp#L142
> > > 
> > > or move translation back to `translate_direct_reduce_op`?
> > 
> > 
> > Because test in tf_lite is failing it won't pass now. Let me know what should I do with it
> 
> please use `translate_prod_op` translator in TFLite since it is common for both FEs. btw, there is an failure in TF layer tests due to accuracy error. Please investigate.
> 
> Best regards, Roman

Probably it is a bug in ARM plugin part because other platforms do not fail. Please use static seed to have the same input data in all platforms. If the failure is reproduced only on ARM, it means CPU plugin for arm part issue and you need just mark this particular case as xfail on arm.

Best regards,
Roman

hub-bla commented:
> > > > Am I allowed to alter `REDUCE_PROD` here
> > > > https://github.com/openvinotoolkit/openvino/blob/102e8754440aec4ff5712bccdff009bf7ee88605/src/frontends/tensorflow_lite/src/op_table.cpp#L142
> > > > 
> > > > or move translation back to `translate_direct_reduce_op`?
> > > 
> > > 
> > > Because test in tf_lite is failing it won't pass now. Let me know what should I do with it
> > 
> > 
> > please use `translate_prod_op` translator in TFLite since it is common for both FEs. btw, there is an failure in TF layer tests due to accuracy error. Please investigate.
> > Best regards, Roman
> 
> Probably it is a bug in ARM plugin part because other platforms do not fail. Please use static seed to have the same input data in all platforms. If the failure is reproduced only on ARM, it means CPU plugin for arm part issue and you need just mark this particular case as xfail on arm.
> 
> Best regards, Roman

I think it was because I moved `REDUCE_PROD` from `translate_direct_reduce_op` to my translator. And left the above line untouched so it was not supported. I reverted that change and it should work now

rkazants commented:
build_jenkins

rkazants commented:
@hub-bla,

please skip the following test case on GPU due to GPU plugins bug:

```
tensorflow_tests/test_tf_ReduceArithmeticOps.py::TestComplexProd::test_reduce[ ie_device:GPU - precision:FP16 - keep_dims:True - params:{'shape': [2, 3, 5], 'axis': 1} ]
```

hub-bla commented:
> @hub-bla,
> 
> please skip the following test case on GPU due to GPU plugin bug:
> 
> ```
> tensorflow_tests/test_tf_ReduceArithmeticOps.py::TestComplexProd::test_reduce[ ie_device:GPU - precision:FP16 - keep_dims:True - params:{'shape': [2, 3, 5], 'axis': 1} ]
> ```

Is this code enough?
```python
if ie_device == 'GPU' and precision == tf.float16 and params['shape'] == [2, 3, 5] and keep_dims==True:
            pytest.skip("GPU plugin bug")
```
Any specific message for the pytest.skip?

rkazants commented on tests/layer_tests/tensorflow_tests/test_tf_ReduceArithmeticOps.py:
```diff
@@ -50,3 +51,51 @@ def test_reduce(self, params, operation, keep_dims, ie_device, precision, ir_ver
+        dict(shape=[3, 1, 2, 4], axis=-2),
+    ]
+
+    @pytest.mark.parametrize("params", test_data)
+    @pytest.mark.parametrize("keep_dims", [True, False])
+    @pytest.mark.nightly
+    @pytest.mark.precommit
+    def test_reduce(self, params, keep_dims, ie_device, precision, ir_version, temp_dir,
+                    use_legacy_frontend):
+        self._test(*self.create_complex_prod_net(**params, keep_dims=keep_dims, ir_version=ir_version,
+                                           use_legacy_frontend=use_legacy_frontend),
+                   ie_device, precision, ir_version, temp_dir=temp_dir,
+                   use_legacy_frontend=use_legacy_frontend)
```
```suggestion
    @pytest.mark.parametrize("params", test_data)
    @pytest.mark.parametrize("keep_dims", [True, False])
    @pytest.mark.nightly
    @pytest.mark.precommit
    def test_reduce(self, params, keep_dims, ie_device, precision, ir_version, temp_dir,
                    use_legacy_frontend):
        if ie_device == 'GPU' and params['shape'] == [2, 3, 5] and keep_dims is True:
            pytest.skip('GPU plugin accuracy error')
        self._test(*self.create_complex_prod_net(**params, keep_dims=keep_dims, ir_version=ir_version,
                                           use_legacy_frontend=use_legacy_frontend),
                   ie_device, precision, ir_version, temp_dir=temp_dir,
                   use_legacy_frontend=use_legacy_frontend)
```

rkazants commented:
> Any specific message for the pytest.skip?

I fixed it. Thanks.

rkazants commented:
build_jenkins

rkazants commented on tests/layer_tests/tensorflow_tests/test_tf_ReduceArithmeticOps.py:
```diff
@@ -50,3 +51,53 @@ def test_reduce(self, params, operation, keep_dims, ie_device, precision, ir_ver
+    @pytest.mark.precommit
+    def test_reduce(self, params, keep_dims, ie_device, precision, ir_version, temp_dir,
+                    use_legacy_frontend):
+        if ie_device == 'GPU' and params['shape'] == [2, 3, 5] and keep_dims is True:
```
```suggestion
        if ie_device == 'GPU' and params['shape'] == [2, 3, 5]:
```

rkazants commented:
build_jenkins

rkazants commented on tests/layer_tests/tensorflow_tests/test_tf_ReduceArithmeticOps.py:
```diff
@@ -50,3 +51,53 @@ def test_reduce(self, params, operation, keep_dims, ie_device, precision, ir_ver
+    @pytest.mark.precommit
+    def test_reduce(self, params, keep_dims, ie_device, precision, ir_version, temp_dir,
+                    use_legacy_frontend):
+        if ie_device == 'GPU' and params['shape'] == [2, 3, 5]:
```
```suggestion
        if ie_device == 'GPU' and params['shape'] in [[2, 3, 5], [3, 1, 2, 4]]:
```

rkazants commented:
build_jenkins
